src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/Utilities/Checksum.Builder.cs

```diff
@@ -2,19 +2,36 @@
 // Licensed under the MIT license. See License.txt in the project root for license information.
 
 using System;
+#if !NET
 using System.Buffers;
+#endif
 using System.Buffers.Binary;
 using System.Diagnostics;
 using System.Runtime.InteropServices;
 using System.Security.Cryptography;
 using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
 
 namespace Microsoft.AspNetCore.Razor.Utilities;
 
 internal sealed partial record Checksum
 {
-    internal readonly ref struct Builder
+    // PERFORMANCE: Care has been taken to avoid using IncrementalHash on .NET Framework, which can cause
+    // threadpool starvation. Essentially, on .NET Framework, IncrementalHash ends up using the OS implementation
+    // of SHA-256, which creates several finalizable objects in the form of safe handles. By creating instances
+    // of SHA256 for .NET Framework (and netstandard2.0), we get the managed code version of SHA-256, which
+    // doesn't have the overhead of using the OS implementations.
+    //
+    // See https://github.com/dotnet/roslyn/issues/67995 for more detail.
+
+    internal readonly ref partial struct Builder
     {
+#if NET
+        private static readonly ObjectPool<IncrementalHash> s_incrementalHashPool = DefaultPool.Create(Policy.Instance);
```

Comment:
What if, instead of these `#ifdef`s around each type of object, we instead do:

```cs
using HashingType =
#if NET5_0_OR_GREATER
    FQN.IncrementalHash;
#else
    FQN.SHA256;
#endif
```

That should remove this ifdef, and `_hash` field one. Not bothered if you don't want to do that refactoring, just a thought.